src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Utf8StringMarshaller.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Marshaller for UTF-8 strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct Utf8StringMarshaller
+    {
+        private byte* allocated;
+        private Span<byte> span;
+
+        // Conversion from a 2-byte 'char' in UTF-16 to bytes in UTF-8 has a maximum of 3 bytes per 'char'
+        // Two bytes ('char') in UTF-16 can be either:
+        //   - Code point in the Basic Multilingual Plane: all 16 bits are that of the code point
+        //   - Part of a pair for a code point in the Supplementary Planes: 10 bits are that of the code point
+        // In UTF-8, 3 bytes are need to represent the code point in first and 4 bytes in the second. Thus, the
+        // maximum number of bytes per 'char' is 3.
+        private const int MaxByteCountPerChar = 3;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf8StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public Utf8StringMarshaller(string str)
+            : this(str, default(Span<byte>))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf8StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public Utf8StringMarshaller(string str, Span<byte> buffer)
+        {
+            allocated = null;
+            span = default;
+            if (str is null)
+                return;
+
+            // + 1 for number of characters in case left over high surrogate is ?
+            // * <MaxByteCountPerChar> (3 for UTF-8)
+            // +1 for null terminator
+            if (buffer.Length >= (str.Length + 1) * MaxByteCountPerChar + 1)
```

Comment:
We don't want to use Encoding.UTF8.GetMaxByteCount(str)?  What happens if this calculation overflows?

Reply:
When it used `GetMaxByteCount`, it seemed to add a bit of overhead (~5% for our tests that were just marshalling a string by value), so I made it match the built-in system that just does the calculation like this. I didn't try it in the macro/asp.net tests, so maybe it doesn't really show up/matter.

Reply:
`StringToCoTaskMemUTF8` is going to call `GetMaxByteCount` anyway. We can inline the logic from `StringToCoTaskMemUTF8` here and check again after calling `GetMaxByteCount` whether the stack buffer is big enough.

Reply:
Also, does this need to allocate the memory using CoTaskMemAlloc? If it is possible, it would be a tiny bit faster to use NativeMemory.

Reply:
For just marshalling by value, it would be possible. For ref, we'd still want CoTaskMemAlloc, since the native side could be expecting that (could transfer ownership / free / replace), so we'd have allocate differently and free accordingly for by value versus ref.